pkg/image/controller/controller.go

```diff
@@ -35,243 +66,59 @@ const retryCount = 2
+//    for kind == DockerImage import or reference underlying image, exact tag (not provided means latest),
 //    for kind != DockerImage reference tag from the same or other ImageStream
-// 3. spec.DockerImageRepository not defined - import tags per its definition.
-// Current behavior of the controller is to process import as far as possible, but
-// we still want to keep backwards compatibility and retries, for that we'll return
-// error in the following cases:
-// 1. connection failure to upstream image repository
-// 2. reading tags when error is different from RepositoryNotFound or RegistryNotFound
-// 3. image retrieving when error is different from RepositoryNotFound, RegistryNotFound or ImageNotFound
-// 4. ImageStreamMapping save error
-// 5. error when marking ImageStream as imported
+//
+// 3. spec.DockerImageRepository not defined - import tags per each definition.
+//
 func (c *ImportController) Next(stream *api.ImageStream) error {
-	if !needsImport(stream) {
+	ok, partial := needsImport(stream)
+	if !ok {
 		return nil
 	}
-	glog.V(4).Infof("Importing stream %s/%s...", stream.Namespace, stream.Name)
+	glog.V(3).Infof("Importing stream %s/%s partial=%t...", stream.Namespace, stream.Name, partial)
 
-	insecure := stream.Annotations[api.InsecureRepositoryAnnotation] == "true"
-	client := c.client
-	if client == nil {
-		client = dockerregistry.NewClient(5 * time.Second)
-	}
-
-	var errlist []error
-	toImport, retry, err := getTags(stream, client, insecure)
-	// return here, only if there is an error and nothing to import
-	if err != nil && len(toImport) == 0 {
-		if retry {
-			return err
-		}
-		return c.done(stream, err.Error(), retryCount)
-	}
-	if err != nil {
-		errlist = append(errlist, err)
-	}
-
-	retry, err = c.importTags(stream, toImport, client, insecure)
-	if err != nil {
-		if retry {
-			return err
-		}
-		errlist = append(errlist, err)
-	}
-
-	if len(errlist) > 0 {
-		return c.done(stream, kerrors.NewAggregate(errlist).Error(), retryCount)
+	isi := &api.ImageStreamImport{
+		ObjectMeta: kapi.ObjectMeta{
+			Name:            stream.Name,
+			Namespace:       stream.Namespace,
+			ResourceVersion: stream.ResourceVersion,
+		},
+		Spec: api.ImageStreamImportSpec{Import: true},
 	}
-
-	return c.done(stream, "", retryCount)
-}
-
-// getTags returns a map of tags to be imported, a flag saying if we should retry
-// imports, meaning not setting the import annotation and an error if one occurs.
-// Tags explicitly defined will overwrite those from default upstream image repository.
-func getTags(stream *api.ImageStream, client dockerregistry.Client, insecure bool) (map[string]api.DockerImageReference, bool, error) {
-	imports := make(map[string]api.DockerImageReference)
-	references := sets.NewString()
-
-	// read explicitly defined tags
-	for tagName, specTag := range stream.Spec.Tags {
-		if specTag.From == nil {
+	for tag, tagRef := range stream.Spec.Tags {
```

Comment:
Also note that partial is not the correct check - tagNeedsImport takes true here (which has different behavior if generation is nil) because we need to handle image streams created that don't have a generation set - in that case we do a single import.